src/execution_plan/optimizations/optimize_cartesian_product.c

```diff
@@ -0,0 +1,127 @@
+*
+* This file is available under the Redis Labs Source Available License Agreement
+*/
+
+#include "optimize_cartesian_product.h"
+#include "../ops/op_filter.h"
+#include "../ops/op_cartesian_product.h"
+#include "optimizations_util.h"
+
+// Collects all consecutive filters beneath given op.
+static OpFilter **_locate_filters(OpBase *cp) {
+	OpBase *parent = cp->parent;
+	OpFilter **filters = array_new(OpFilter *, 0);
+
+	while(parent && parent->type == OPType_FILTER) {
+		OpFilter *filter_op = (OpFilter *)parent;
+		filters = array_append(filters, filter_op);
+		parent = parent->parent;
+	}
+
+	return filters;
+}
+
+/**
+ * @brief  Creates a structure of dual branched catesian product, followed by a filter.
+ *         The generated record from the cartesian product completely satisfies the filter entities.
+ * @param  *plan: Execution plan to associate its mapping with the new CP operation.
+ * @param  *lhs: Left side branch.
+ * @param  *rhs: Right side branch.
+ * @param  *filter: Trailing filter operation.
+ * @retval Pointer to the filter operation.
+ */
+static OpBase *_chain_cp_and_filter(const ExecutionPlan *plan, OpBase *lhs, OpBase *rhs,
+									OpBase *filter) {
+	OpBase *cp = NewCartesianProductOp(plan);
+	ExecutionPlan_AddOp(cp, lhs);
+	ExecutionPlan_AddOp(cp, rhs);
+	ExecutionPlan_AddOp(filter, cp);
+	return filter;
+}
+
+static void _optimize_cartesian_product(ExecutionPlan *plan, OpBase *cp) {
+	// Retrieve all filter operations located upstream from the Cartesian Product.
+	OpFilter **filter_ops = _locate_filters(cp);
+	uint filter_count = array_len(filter_ops);
+
+	// For each stream joined by the Cartesian product, collect all entities the stream resolves.
+	int stream_count = cp->childCount;
+	rax *stream_entities[stream_count];
+	OptimizeUtils_BuildStreamFromOp(cp, stream_entities, stream_count);
+
+	for(uint i = 0; i < filter_count; i ++) {
+		// Try to create a new dual-branched cartesian product, followed by the current filter.
+		OpFilter *filter_op = filter_ops[i];
+
+		/* A new cartesian product can be generated if both sides of the filter can be fully and
+		 * separately resolved by exactly two child streams. */
+		FT_FilterNode *f = filter_op->filterTree;
+
+		/* Make sure LHS of the filter is resolved by a stream. */
+		AR_ExpNode *lhs = f->pred.lhs;
+		uint lhs_resolving_stream = OptimizeUtils_RelateExpToStream(lhs, stream_entities, stream_count);
+		if(lhs_resolving_stream == NOT_RESOLVED) continue;
+		/* Make sure RHS of the filter is resolved by a stream. */
+		AR_ExpNode *rhs = f->pred.rhs;
+		uint rhs_resolving_stream = OptimizeUtils_RelateExpToStream(rhs, stream_entities, stream_count);
+		if(rhs_resolving_stream == NOT_RESOLVED) continue;
+
+		// This stream is solved by a single cartesian product child and needs to be propagated up.
+		if(lhs_resolving_stream == rhs_resolving_stream) {
+			OptimizeUtils_MigrateFilterOp(plan, cp->children[rhs_resolving_stream], filter_op);
```

Comment:
Unnecessary new function, use a combination of `ExecutionPlan_RemoveOp` and `ExecutionPlan_PushBelow`
I prefer not to introduce new functions but to work with what we have, otherwise we'll be overwhelmed with functions.

Reply:
we use this method 4 times, in two different files. I think we shold avoid code duplication

Reply:
removed